xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -48,7 +50,7 @@ static Builder newBuilder() {
     }
 
     static final class Builder {
-      private ConfigUpdate base;
+      private final ConfigUpdate base;
```

Comment:
With this `builder.build()` is returning the same instance `base` and that does't look right.  Ideally your builder should save all the field values and construct a new instance in `build()` with those values. Instead you have a single `final` instance `base` and non-final instance variables like `clusterName` in `ConfigUpdate`.

Reply:
`ConfigUpdate` is immutable, its fields are set through a `ConfigUpdate.Builder` instance. Every time you call `ConfigUpdate.newBuilder()` you are creating a new `ConfigUpdate` instance inside. `ConfigUpdate.Builder#build()` returns it to you. This is a boilerplate pattern, I don't see anything is wrong here.

Reply:
The boilerplate pattern is for `Builder.build()` to return a new instance every single time. Here you are building a `ConfigUpdate` just once in the `Builder` constructor and just returning that instance in `build()`. If somebody uses the same builder instance and sets a different cluster-name (by calling `setClusterName(String clusterName)`) you will just be changing the clusterName of the previous instance and returning the same instance. Ideally you should have `clusterName` final in `ConfigUpdate` but `base` non-final in `ConfigUpdate.Builder`. It may not matter in this case (if clients always use newBuilder) but the pattern is misleading.

Reply:
Huh, I see what you mean.

Reply:
I agree with @sanjaypujare for I think we should follow suit the pattern of generated code of protobuf message (not protobuf lite), which returns a new instance each time when `build()`. Just to avoid different patterns.